packages/react/src/useTableUtils/helpers.ts

```diff
@@ -145,82 +130,202 @@ export const getTableColumns = (spec: TableSpec) => {
-          } else {
-            row[columnApiIdentifier] = record[columnApiIdentifier]?.edges.map((edge: any) => edge.node[relatedField.apiIdentifier]);
-          }
-        }
+      row[targetColumn.name] = targetColumn.render({ record });
+      continue;
+    }
+
+    const { field: columnPath } = getCellDetailColumnByColumnValue(targetColumn);
+    const { firstPathSegment, targetField, isHasMany, isHasOneOrBelongsTo } = getFieldInformationByColumnPath(
+      spec.fieldMetadataTree,
+      columnPath
+    );
+
+    // The column path doesn't specify a related field, so we need to get the default display field
+    if (isHasOneOrBelongsTo || isHasMany) {
+      const relatedField = maybeGetDefaultDisplayFieldMetadata(spec.fieldMetadataTree, columnPath) ?? targetField;
+      if (isHasMany) {
+        row[columnPath] = get(record, firstPathSegment)?.edges.map((edge: any) => edge.node[relatedField.apiIdentifier]);
       } else {
-        row[columnApiIdentifier] = record[columnApiIdentifier];
+        row[columnPath] = get(record, firstPathSegment)?.[relatedField.apiIdentifier];
       }
+    } else {
+      row[columnPath] = get(record, columnPath);
     }
   }
 
   return row;
 };
+
+const getFieldInformationByColumnPath = (fieldMetadataTree: TableSpec["fieldMetadataTree"], columnPath: string) => {
+  const firstPathSegment = columnPath.split(".")[0];
+  const firstField = getFieldMetadataByColumnPath(fieldMetadataTree, firstPathSegment);
+  const targetField = getFieldMetadataByColumnPath(fieldMetadataTree, columnPath);
+
+  const isHasOneOrBelongsTo = isHasOneOrBelongsToField(firstField);
+  const isHasMany = isHasManyField(firstField);
+
+  return {
+    firstPathSegment,
+    firstField,
+    targetField,
+
+    isHasOneOrBelongsTo,
+    isHasMany,
+  };
+};
+
+const isCustomCellColumn = (value: any): value is CustomCellColumn => {
+  return typeof value === "object" && value !== null && "name" in value && "render" in value;
+};
+
+const getCellDetailColumnByColumnValue = (column: string | CellDetailColumn): CellDetailColumn => {
+  return typeof column === "string" ? { field: column } : column;
+};
+
+const isColumnSortable = (fieldMetadata: FieldMetadataFragment, sortable: boolean | undefined) => {
+  return sortable ?? ("sortable" in fieldMetadata && fieldMetadata.sortable);
+};
+
+const mergeColumnPathByFieldType = (columnPath: string, newSegment: string, field: { fieldType: GadgetFieldType }) => {
+  if (field.fieldType === GadgetFieldType.HasMany) {
+    return `${columnPath}.edges.node.${newSegment}`;
+  }
+
+  return `${columnPath}.${newSegment}`;
+};
+
+const isHasOneOrBelongsToField = (field: FieldMetadataFragment) => {
+  return field.fieldType === GadgetFieldType.HasOne || field.fieldType === GadgetFieldType.BelongsTo;
+};
+
+const isHasManyField = (field: FieldMetadataFragment) => {
+  return field.fieldType === GadgetFieldType.HasMany;
+};
```

Comment:
There are a couple of places throughout this file where you're doing these checks again instead of using these helpers

Reply:
Ahh right, updated to use these functions instead.